libexec/rmail/rmail.8
.\" Copyright (c) 1983, 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)rmail.8	6.8 (Berkeley) 4/20/91
.\"
.\"	$Header: /usr/tmp/cvs2git/cvsroot-netbsd/src/libexec/rmail/Attic/rmail.8,v 1.3 1993/03/23 00:27:15 cgd Exp $
.\"
.Vx
.Vx
.Dd April 20, 1991
.Dt RMAIL 8
.Os BSD 4.2
.Sh NAME
.Nm rmail
.Nd handle remote mail received via uucp
.Sh SYNOPSIS
.Nm rmail
.Ar user ...
.Sh DESCRIPTION
.Nm Rmail
interprets incoming mail received via
.Xr uucp 1 ,
collapsing ``From'' lines in the form generated
by
.Xr binmail  1
into a single line of the form ``return-path!sender'',
and passing the processed mail on to
.Xr sendmail  8  .
.Pp
.Nm Rmail
is explicitly designed for use with
.Xr uucp
and
.Xr sendmail  .
.Sh SEE ALSO
.Xr binmail 1 ,
.Xr uucp 1 ,
.Xr sendmail 8
.Sh HISTORY
.Nm Rmail
apeared in 4.2 BSD.
.Sh BUGS
.Nm Rmail
should not reside in /bin.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3291 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	riscv: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	vax: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]
	gpt(8): Add various riscv-specific partition types. [thorpej 20251011]
	fsck_lfs(8): Enable checking LFS64 volumes. [perseant 20251012]
	m68k: New port: mac68ksf - Macintosh 68k softfloat. [nat 20251020]
	sc16is7xx(4): NXP SC16IX7xx family of UART chips that are behind
		a I2C or SPI bus.  [brad 20251024]
	compat_linux(8): Add support for semtimedop(2), from Ricardo Branco.
		[christos 20251026]
	ftp(1): Add "connect" as a synonym for "open". [lukem 20251028]
	ia64: Switch to GCC 14. [mrg 20251029]
	lfs: Add in-kernel cleaner. [perseant 20251106]
	m68k: Experimental new pmap module for systems with 68k MMUs
		(plus the HP MMU). [thorpej 20251108]
	compat_linux(8): Add prctl(2) lwp_{g,s}etname, from Kevin Bloom.
		[christos 20251110]
	sparc64: Add implementation of Sun's VIS instruction API.
		[nia 20251113]
	iconv(1): Support -o outfile like gnu does and check writes for
		failure. [christos 20251121]
	cgd(4): Improve throughput of AES disk encryption using SSE2.
		[riastradh 20251123]
	powerpc: Switch to GCC 14. [martin 20251128]
	dreamcast, atu(4), icp(4), gpsa(8): Relicense various NetBSD
		components from 4-clause to 3-clause BSD. [nia 20251128]
	umcpmio(4): Reorg and support for the MCP2210, USB to SPI
		bridge. [brad 20251129]
	ucom(4): Allow using as boot console. [manu 20251129]
	sparc64: Use 64-bit optimized AES code for disk encryption.
		[nia 20251201]
	iwm(4): Update driver code, improving performance and stability.
		[mlelstv 20251201]
	lfs: Various stability improvements. [perseant 20251202]
	viaide(4): Identify and attach VT6420 IDE (PATA) controller.
		[andvar 20251204]
	diff(1), diff3(1): Import BSD-licensed diff to replace GNU
		diffutils, build with MKBSDDIFF=yes. [nia 20251204]
	evbppc: Enable PaX security features in WII kernel configuration.
		[nia 20251204]
	amd64: Support ramdisks when booting in GENPVH mode.
		[khorben 20251205]
	ftp(1): Fix ASCII transfers with progress bar. [lukem 20251206]
	evbppc: wii: Add support for GameCube controller sockets
		as uhid(4) (now "userspace hid") devices.
		[jmcneill 20251208]
	mips: Switch to GCC 14. [mrg 20251209]
	openresolv(8): import 3.17.1 (patched to 3.17.3) [roy 20251213]
	editline(3): Add a method to set custom word characters,
		from bsiegert. [christos 20251214]
	evbarm: Add support for TI OMAP4 PandaBoard-ES, from Rui-Xiang Guo.
		[skrll 20251216]
	tzcode: Updated to 2025c. [christos 20251218]
	tzdata: Updated to 2025c (using 2025cgtz) [kre 20251219]
	re(4): Made the code MPSAFE with a private mutex.
		[mlelstv 20251220]
	lm(4): Support five aux fan sensors for NCT6794D. [isaki 20251224]
	inetd(8): Add an optional accept limit to stream/nowait services.
		[mlelstv 20251227]
	sun2: Add various MK build knobs and disable them on sun2 to trim
		down the installation. [thorpej 20251228]
	diff(1): Switch the default to MKBSDDIFF=yes. [nia 20260102]
	i386, amd64, sparc64: Exclude various base features from the
		CD-ROM installer and add checks that cause ISO image

libexec/rmail/rmail.c
/*
 * Copyright (c) 1981, 1988 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1981, 1988 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)rmail.c	4.15 (Berkeley) 5/31/90";
static char rcsid[] = "$Header: /usr/tmp/cvs2git/cvsroot-netbsd/src/libexec/rmail/Attic/rmail.c,v 1.3 1993/03/23 00:27:17 cgd Exp $";
#endif /* not lint */

/*
 * RMAIL -- UUCP mail server.
 *
 *	This program reads the >From ... remote from ... lines that
 *	UUCP is so fond of and turns them into something reasonable.
 *	It calls sendmail giving it a -f option built from these lines. 
 */

#include <sysexits.h>
#include <sys/types.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <stdio.h>
#include <paths.h>

typedef char bool;
#define TRUE	1
#define FALSE	0

extern char *index();
extern char *rindex();

char *Domain = "UUCP";		/* Default "Domain" */

main(argc, argv)
	int argc;
	char **argv;
{
	char lbuf[1024];	/* one line of the message */
	char from[512];		/* accumulated path of sender */
	char ufrom[512];	/* user on remote system */
	char sys[512];		/* a system in path */
	char fsys[512];		/* first system in path */
	char junk[1024];	/* scratchpad */
	char *args[100];	/* arguments to mailer command */
	register char *cp;
	register char *uf = NULL;	/* ptr into ufrom */
	int i;
	long position;
	struct stat sbuf;
#ifdef DEBUG
	bool Debug;

	if (argc > 1 && strcmp(argv[1], "-T") == 0) {
		Debug = TRUE;
		argc--;
		argv++;
	}
#endif

	if (argc < 2) {
		fprintf(stderr, "Usage: rmail user ...\n");
		exit(EX_USAGE);
	}
	if (argc > 2 && strncmp(argv[1], "-D", 2) == 0) {
		Domain = &argv[1][2];
		argc -= 2;
		argv += 2;
	}
	from[0] = '\0';
	fsys[0] = '\0';
	(void) strcpy(ufrom, _PATH_DEVNULL);

	for (position = 0;; position = ftell(stdin)) {
		if (fgets(lbuf, sizeof lbuf, stdin) == NULL)
			exit(EX_DATAERR);
		if (strncmp(lbuf, "From ", 5) != 0 &&
		    strncmp(lbuf, ">From ", 6) != 0)
			break;
		(void) sscanf(lbuf, "%s %s", junk, ufrom);
		cp = lbuf;
		uf = ufrom;
		for (;;) {
			cp = index(cp + 1, 'r');
			if (cp == NULL) {
				register char *p = rindex(uf, '!');

				if (p != NULL) {
					*p = '\0';
					(void) strcpy(sys, uf);
					uf = p + 1;
					break;
				}
				(void) strcpy(sys, "");
				break;	/* no "remote from" found */
			}
#ifdef DEBUG
			if (Debug)
				printf("cp='%s'\n", cp);
#endif
			if (strncmp(cp, "remote from ", 12) == 0)
				break;
		}
		if (cp != NULL)
			(void) sscanf(cp, "remote from %s", sys);
		if (fsys[0] == '\0')
			(void) strcpy(fsys, sys);
		if (sys[0]) {
			(void) strcat(from, sys);
			(void) strcat(from, "!");
		}
#ifdef DEBUG
		if (Debug)
			printf("ufrom='%s', sys='%s', from now '%s'\n", uf, sys, from);
#endif
	}
	if (uf == NULL) {	/* No From line was provided */
		fprintf(stderr, "No From line in rmail\n");
		exit(EX_DATAERR);
	}
	(void) strcat(from, uf);
	(void) fstat(0, &sbuf);
	(void) lseek(0, position, L_SET);

	/*
	 * Now we rebuild the argument list and chain to sendmail. Note that
	 * the above lseek might fail on irregular files, but we check for
	 * that case below. 
	 */
	i = 0;
	args[i++] = _PATH_SENDMAIL;
	args[i++] = "-oee";		/* no errors, just status */
	args[i++] = "-odq";		/* queue it, don't try to deliver */
	args[i++] = "-oi";		/* ignore '.' on a line by itself */
	if (fsys[0] != '\0') {		/* set sender's host name */
		static char junk2[512];

		if (index(fsys, '.') == NULL) {
			(void) strcat(fsys, ".");
			(void) strcat(fsys, Domain);
		}
		(void) sprintf(junk2, "-oMs%s", fsys);
		args[i++] = junk2;
	}
					/* set protocol used */
	(void) sprintf(junk, "-oMr%s", Domain);
	args[i++] = junk;
	if (from[0] != '\0') {		/* set name of ``from'' person */
		static char junk2[512];

		(void) sprintf(junk2, "-f%s", from);
		args[i++] = junk2;
	}
	for (; *++argv != NULL; i++) {
		/*
		 * don't copy arguments beginning with - as they will
		 * be passed to sendmail and could be interpreted as flags
		 * should be fixed in sendmail by using getopt(3), and
		 * just passing "--" before regular args.
		 */
		if (**argv != '-')
			args[i] = *argv;
	}
	args[i] = NULL;
#ifdef DEBUG
	if (Debug) {
		printf("Command:");
		for (i = 0; args[i]; i++)
			printf(" %s", args[i]);
		printf("\n");
	}
#endif
	if ((sbuf.st_mode & S_IFMT) != S_IFREG) {
		/*
		 * If we were not called with standard input on a regular
		 * file, then we have to fork another process to send the
		 * first line down the pipe. 
		 */
		int pipefd[2];
#ifdef DEBUG
		if (Debug)
			printf("Not a regular file!\n");
#endif
		if (pipe(pipefd) < 0)
			exit(EX_OSERR);
		if (fork() == 0) {
			/*
			 * Child: send the message down the pipe. 
			 */
			FILE *out;

			out = fdopen(pipefd[1], "w");
			close(pipefd[0]);
			fputs(lbuf, out);
			while (fgets(lbuf, sizeof lbuf, stdin))
				fputs(lbuf, out);
			(void) fclose(out);
			exit(EX_OK);
		}
		/*
		 * Parent: call sendmail with pipe as standard input 
		 */
		close(pipefd[1]);
		dup2(pipefd[0], 0);
	}
	execv(_PATH_SENDMAIL, args);
	fprintf(stderr, "Exec of %s failed!\n", _PATH_SENDMAIL);
	exit(EX_OSERR);
}
